subsys/bluetooth/mesh/access.c

```diff
@@ -2614,3 +2634,220 @@ uint8_t bt_mesh_comp_parse_page(struct net_buf_simple *buf)
 
 	return page;
 }
+
+#if defined CONFIG_BT_MESH_ACCESS_TX_MSG_RANDOMIZER
+
+struct access_message_chunk {
+	sys_snode_t node;
+	uint8_t data[CONFIG_BT_MESH_DELAYED_ACCESS_MSG_CHUNK_SIZE];
+};
+
+static void randomizer_handler(struct k_work *w);
+
+static struct access_message {
+	bool busy;
+	struct bt_mesh_msg_ctx ctx;
+	const struct bt_mesh_model *model;
+	const struct bt_mesh_send_cb *cb;
+	void *cb_data;
+	uint32_t fired_time;
+	size_t len;
+	sys_slist_t chunks;
+} access_messages[CONFIG_BT_MESH_DELAYED_ACCESS_MSG_COUNT];
+
+static struct {
+	struct k_work_delayable random_delay;
+	struct access_message *pending_msg;
+} access_randomizer = {.random_delay = Z_WORK_DELAYABLE_INITIALIZER(randomizer_handler)};
+
+K_MEM_SLAB_DEFINE_STATIC(access_message_chunks, sizeof(struct access_message_chunk),
+			 CONFIG_BT_MESH_DELAYED_ACCESS_MSG_CHUNK_COUNT, 4);
+
+static void reschedule_randomizer(void)
+{
+	uint32_t fired_time_min = 0;
+	uint32_t curr_time;
+	k_timeout_t delay = K_NO_WAIT;
+
+	access_randomizer.pending_msg = NULL;
+
+	for (int i = 0; i < CONFIG_BT_MESH_DELAYED_ACCESS_MSG_COUNT; i++) {
+		if (access_messages[i].busy) {
+			if (fired_time_min < access_messages[i].fired_time) {
+				fired_time_min = access_messages[i].fired_time;
+				access_randomizer.pending_msg = &access_messages[i];
+			}
+		}
+	}
+
+	if (!access_randomizer.pending_msg) {
+		return;
+	}
+
+	curr_time = k_uptime_get_32();
+	if (curr_time < fired_time_min) {
+		delay = K_MSEC(fired_time_min - curr_time);
+	}
+
+	k_work_reschedule(&access_randomizer.random_delay, delay);
+}
+
+static int allocate_msg_chunks(struct access_message *msg, int number)
+{
+	struct access_message_chunk *chunk;
+	int err;
+
+	for (int i = 0; i < number; i++) {
+		err = k_mem_slab_alloc(&access_message_chunks, (void **)&chunk, K_NO_WAIT);
+		if (err) {
+			LOG_WRN("Unable allocate %u chunks, allocated %u", number, i);
+			return i;
+		}
+		sys_slist_append(&msg->chunks, &chunk->node);
+	}
```

Comment:
Have you considered [Memory Blocks Allocator](https://docs.zephyrproject.org/latest/kernel/memory_management/sys_mem_blocks.html)? It allows it allocate required number of blocks in a [single call](https://docs.zephyrproject.org/latest/kernel/memory_management/sys_mem_blocks.html#allocating-memory-blocks). We may then not need (unless I miss something) linked list for each block (`struct access_message_chunk`), but keep an array of pointers with size set to `BT_MESH_TX_SEG_MAX` in `struct access_message`:
```
struct access_message {
...
        uintptr segs[CONFIG_BT_MESH_TX_SEG_MAX];
...
} access_messages[...] = ...
```
Releasing chunks is also [simple](https://docs.zephyrproject.org/latest/kernel/memory_management/sys_mem_blocks.html#releasing-a-memory-block).

Reply:
Yep, code with memory block becomes simpler but memory consumption grows:

```
feature based on slab allocator
Memory region         Used Size  Region Size  %age Used
           FLASH:      213228 B         1 MB     20.34%
             RAM:       40270 B       256 KB     15.36%
        IDT_LIST:          0 GB         2 KB      0.00%

feature based on memory block
Memory region         Used Size  Region Size  %age Used
           FLASH:      214000 B         1 MB     20.41%
             RAM:       40254 B       256 KB     15.36%
        IDT_LIST:          0 GB         2 KB      0.00%
```

flash diff is 772 bytes. I'd suggest leaving slab usage.